src/openforms/formio/dynamic_config/date.py

```diff
@@ -0,0 +1,143 @@
+import operator
+from datetime import datetime, time
+from typing import Literal, Optional, TypedDict, Union, cast
+
+from django.utils import timezone
+from django.utils.translation import gettext_lazy as _
+
+from dateutil.relativedelta import relativedelta
+from glom import assign, glom
+
+from openforms.typing import DataMapping
+
+from ..typing import Component
+from .registry import BasePlugin, register
+
+NOW_VARIABLE = "now"
+
+
+class DateConstraintDelta(TypedDict):
+    years: Optional[int]
+    months: Optional[int]
+    days: Optional[int]
+
+
+class DateConstraintConfiguration(TypedDict):
+    mode: Literal["fixedValue", "future", "past", "relativeToVariable"]
+    includeToday: Optional[bool]
+    variable: Optional[str]
+    delta: Optional[DateConstraintDelta]
+    operator: Optional[Literal["add", "subtract"]]
+
+
+class OpenFormsConfig(TypedDict):
+    minDate: Optional[DateConstraintConfiguration]
+    maxDate: Optional[DateConstraintConfiguration]
+
+
+class DatePickerConfig(TypedDict):
+    minDate: Optional[str]
+    maxDate: Optional[str]
+
+
+class FormioDateComponent(Component):
+    openForms: Optional[OpenFormsConfig]
+    datePicker: Optional[DatePickerConfig]
+
+
+DatetimeLike = Optional[Union[datetime, str]]
+
+
+@register("date")
+class DateComponent(BasePlugin):
+    """
+    Implement behaviour for our custom date component
+
+    In the JS, this component type inherits from Formio datetime component. See
+    ``src/openforms/js/components/form/date.js`` for the various configurable options.
+    """
+
+    verbose_name = _("Date component")
+
+    @classmethod
+    def mutate(cls, component: FormioDateComponent, data: DataMapping) -> None:
+        for key in ("minDate", "maxDate"):
+            config = cast(
+                DateConstraintConfiguration,
+                glom(component, f"openForms.{key}", default=None),
+            )
+            if config is None:
+                continue
+
+            if (mode := config.get("mode")) is None:
+                continue
+
+            # formio has set the datePicker.{key} value
+            if mode == "fixedValue":
```

Comment:
I see a bunch of.magic strings: fixed value, future, past, add, subtract, relativeToVariable, includeToday. Maybe define constants (like the now-variable).

Reply:
Hrm, the `now` variable constant is done because it's external to this module and importing the actual variable from a different app felt too much like violating separation of concerns.

The "magic strings" are type-checked in editors (and with mypy once we get that set up) through the type declaration in the `TypedDict` at https://github.com/open-formulieren/open-forms/pull/2026/files#diff-19dff21508a2142d75a9ac65d2293e0f4cff5acc1506d0544ec04c261ef558dfR26 - this effectively specifies the allowed collection of values for `mode`.

I just tried out defining it as a constant and using that constant in the type declaration like `mode: Literal[MODE_FIXED_VALUE, "future", "past", "relativeToVariable"]`, but that's not valid python - they must all be literals.